File: packages/inputlayer-py/src/inputlayer/result.py
"""ResultSet - typed, iterable query results."""

from __future__ import annotations

from dataclasses import dataclass, field
from types import SimpleNamespace
from typing import Any, Iterator


@dataclass
class ResultSet:
    """Container for query results.

    Supports iteration, indexing, and conversion to dicts/tuples/DataFrames.
    """

    columns: list[str]
    rows: list[list[Any]]
    row_count: int = 0
    total_count: int = 0
    truncated: bool = False
    execution_time_ms: int = 0
    row_provenance: list[str] | None = None
    has_ephemeral: bool = False
    ephemeral_sources: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    # Optional Relation class for typed iteration
    _relation_cls: type | None = field(default=None, repr=False)

    def __post_init__(self) -> None:
        if self.row_count == 0:
            self.row_count = len(self.rows)
        if self.total_count == 0:
            self.total_count = self.row_count

    def __len__(self) -> int:
        return self.row_count

    def __bool__(self) -> bool:
        return self.row_count > 0

    def __iter__(self) -> Iterator[Any]:
        for row in self.rows:
            yield self._row_to_obj(row)

    def __getitem__(self, idx: int) -> Any:
        return self._row_to_obj(self.rows[idx])

    def first(self) -> Any | None:
        """Return the first row as an object, or None if empty."""
        if not self.rows:
            return None
        return self._row_to_obj(self.rows[0])

    def scalar(self) -> Any:
        """Return the single value from a 1×1 result."""
        if not self.rows or not self.rows[0]:
            raise ValueError("No results to extract scalar from")
        return self.rows[0][0]

    def to_dicts(self) -> list[dict[str, Any]]:
        """Convert all rows to list of dicts."""
        return [dict(zip(self.columns, row)) for row in self.rows]

    def to_tuples(self) -> list[tuple[Any, ...]]:
        """Convert all rows to list of tuples."""
        return [tuple(row) for row in self.rows]

    def to_df(self) -> Any:
        """Convert to a pandas DataFrame. Requires pandas."""
        try:
            import pandas as pd
        except ImportError:
            raise ImportError(
                "pandas is required for to_df(). "
                "Install with: pip install inputlayer[pandas]"
            )
        return pd.DataFrame(self.rows, columns=self.columns)

    def _row_to_obj(self, row: list[Any]) -> Any:
        """Convert a row to a typed object or SimpleNamespace."""
        if self._relation_cls is not None:
            try:
                kwargs = dict(zip(self.columns, row))
                return self._relation_cls(**kwargs)
            except Exception:
                pass
        return SimpleNamespace(**dict(zip(self.columns, row)))

File: packages/inputlayer-js/src/result.ts
/**
 * ResultSet - typed, iterable query results.
 */

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyRow = any[];

export interface ResultSetOptions {
  columns: string[];
  rows: AnyRow[];
  rowCount?: number;
  totalCount?: number;
  truncated?: boolean;
  executionTimeMs?: number;
  rowProvenance?: string[];
  hasEphemeral?: boolean;
  ephemeralSources?: string[];
  warnings?: string[];
}

/**
 * Container for query results.
 * Supports iteration, indexing, and conversion to various formats.
 */
export class ResultSet implements Iterable<Record<string, unknown>> {
  readonly columns: string[];
  readonly rows: AnyRow[];
  readonly rowCount: number;
  readonly totalCount: number;
  readonly truncated: boolean;
  readonly executionTimeMs: number;
  readonly rowProvenance?: string[];
  hasEphemeral: boolean;
  ephemeralSources: string[];
  warnings: string[];

  constructor(opts: ResultSetOptions) {
    this.columns = opts.columns;
    this.rows = opts.rows;
    this.rowCount = opts.rowCount ?? opts.rows.length;
    this.totalCount = opts.totalCount ?? this.rowCount;
    this.truncated = opts.truncated ?? false;
    this.executionTimeMs = opts.executionTimeMs ?? 0;
    this.rowProvenance = opts.rowProvenance;
    this.hasEphemeral = opts.hasEphemeral ?? false;
    this.ephemeralSources = opts.ephemeralSources ?? [];
    this.warnings = opts.warnings ?? [];
  }

  /** Number of result rows. */
  get length(): number {
    return this.rowCount;
  }

  /** True if there are any results. */
  get isEmpty(): boolean {
    return this.rowCount === 0;
  }

  /** Get a single row as a keyed object. */
  get(index: number): Record<string, unknown> {
    return this.rowToObj(this.rows[index]);
  }

  /** Get the first row or undefined if empty. */
  first(): Record<string, unknown> | undefined {
    if (this.rows.length === 0) return undefined;
    return this.rowToObj(this.rows[0]);
  }

  /** Return the single value from a 1x1 result. */
  scalar(): unknown {
    if (this.rows.length === 0 || this.rows[0].length === 0) {
      throw new Error('No results to extract scalar from');
    }
    return this.rows[0][0];
  }

  /** Convert all rows to a list of keyed objects. */
  toDicts(): Array<Record<string, unknown>> {
    return this.rows.map((row) => this.rowToObj(row));
  }

  /** Convert all rows to a list of tuples (arrays). */
  toTuples(): AnyRow[] {
    return this.rows.map((row) => [...row]);
  }

  /** Iterate over rows as keyed objects. */
  [Symbol.iterator](): Iterator<Record<string, unknown>> {
    let i = 0;
    const self = this;
    return {
      next(): IteratorResult<Record<string, unknown>> {
        if (i < self.rows.length) {
          return { value: self.rowToObj(self.rows[i++]), done: false };
        }
        return { value: undefined, done: true };
      },
    };
  }

  private rowToObj(row: AnyRow): Record<string, unknown> {
    const obj: Record<string, unknown> = {};
    for (let i = 0; i < this.columns.length && i < row.length; i++) {
      obj[this.columns[i]] = row[i];
    }
    return obj;
  }
}
